esp_linenoise/include/esp_linenoise.h

```diff
@@ -0,0 +1,303 @@
+ * @param bold Pointer to an integer to set bold attribute (non-zero for bold).
+ * @param user_ctx User-defined context pointer passed from the configuration.
+ * @return A dynamically allocated hint string, or NULL if no hint is available.
+ */
+typedef char *(*esp_linenoise_hints_callback)(const char *str, int *color, int *bold, void *user_ctx);
+
+/**
+ * @brief Callback function to free hint strings returned by the hints callback.
+ *
+ * @param ptr Pointer to the hint string to be freed.
+ * @param user_ctx User-defined context pointer passed from the configuration.
+ */
+typedef void (*esp_linenoise_free_hints_callback)(void *ptr, void *user_ctx);
+
+/**
+ * @brief Function pointer type for reading bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer to store the read bytes.
+ * @param count Number of bytes to read.
+ * @param user_ctx User-defined context pointer passed from the configuration.
+ * @return Number of bytes read, or -1 on error.
+ */
+typedef ssize_t (*esp_linenoise_read_bytes_fn)(int fd, void *buf, size_t count, void *user_cx);
+
+/**
+ * @brief Function pointer type for writing bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer containing bytes to write.
+ * @param count Number of bytes to write.
+ * @param user_ctx User-defined context pointer passed from the configuration.
+ * @return Number of bytes written, or -1 on error.
+ */
+typedef ssize_t (*esp_linenoise_write_bytes_fn)(int fd, const void *buf, size_t count, void *user_ctx);
+
+/**
+ * @brief Structure defining the parameters needed by a linenoise
+ * instance
+ */
+typedef struct esp_linenoise_config {
+    const char *prompt; /*!< Prompt string displayed to the user */
+    size_t max_cmd_line_length; /*!< Maximum length (in bytes) of the input command line */
+    int history_max_length; /*!< Maximum number of entries to store in command history */
+    int in_fd; /*!< File descriptor to read input from (e.g., STDIN_FILENO) */
+    int out_fd; /*!< File descriptor to write output to (e.g., STDOUT_FILENO) */
+    bool allow_multi_line; /*!< Whether to allow multi-line input (true to enable) */
+    bool allow_empty_line; /*!< Whether to allow accepting an empty line as valid input */
+    bool allow_dumb_mode; /*!< Whether to allow running in dumb terminal mode (without advanced features) */
+    esp_linenoise_completion_callback completion_cb; /*!< Callback function for handling input completions */
+    esp_linenoise_hints_callback hints_cb; /*!< Callback function to provide input hints (e.g., usage suggestions) */
+    esp_linenoise_free_hints_callback free_hints_cb; /*!< Callback function to free hints returned by `hints_cb` */
+    void *user_ctx; /*!< User context information passed to the callbacks */
+    esp_linenoise_read_bytes_fn read_bytes_fn; /*!< Function used to read bytes from the input stream */
+    esp_linenoise_write_bytes_fn write_bytes_fn; /*!< Function used to write bytes to the output stream */
```

Comment:
Why are some callback names end with `_callback` and others with `_fn`? Can we make them consistent?

I would also recommend keeping `_t` suffix for function pointer typedefs.